src/bootstrap/compile.rs

```diff
@@ -1163,6 +1163,33 @@ pub fn run_cargo(builder: &Builder, cargo: &mut Command, stamp: &Path, is_check:
 
             toplevel.push((file_stem, extension, expected_len));
         }
+        for filename in filenames {
+            // Skip files like executables
+            if !filename.ends_with(".rlib") &&
+               !filename.ends_with(".lib") &&
+               !is_dylib(&filename) &&
+               !(is_check && filename.ends_with(".rmeta")) {
+                break;
+            }
+
+            let filename = Path::new(&*filename);
+
+            // If this was an output file in the "host dir" we don't actually
+            // worry about it, it's not relevant for us.
+            if filename.starts_with(&host_root_dir) {
+                let f = filename.file_name();
+                if deps.iter().find(|v| v.file_name() == f).is_none() {
+                    deps.push(filename.to_path_buf());
+                }
+                continue;
+            }
+
+            // If this was output in the `deps` dir then this is a precise file
+            // name (hash included) so we start tracking it.
+            if filename.starts_with(&target_deps_dir) {
```

Comment:
Update this comment to state something along the lines of "We've already added this in the previous loop iteration"